avocado/core/task/statemachine.py

```diff
@@ -218,6 +218,15 @@ async def triage(self):
                 )
                 return
         if runtime_task.task.category != "test":
+            # generate post plugins tasks after the test finish
+            if type(runtime_task) is PostRuntimeTaskPrototype:
+                post_tasks = runtime_task.get_post_plugin_tasks()
```

Comment:
I had the same concerns as you about doing the creation “this late”. But I found a bigger problem than this, and I didn't find a better way how to solve it. The behavior of post plugin might change base on the test result. We have to support plugins which will run only when the test ends with specific results (pass, fail, skip etc.) like “post test hooks for cleaning up the enviroment which can be used by avocado-vt”. Because of this problem, we can't create a post-test task before we know the test results. That is why I used this solution.

I can see another option how to solve this and create the post-test task sooner. We will create all post-test task without checking the test results (create tasks for test “pass” also tasks for test “fail”). Then we spawn those tasks and the runner will decide base on the test result what to do with the task (it can run, or it can stop). But it is not an optimized solution, and it will make creation of new runners much harder. That is why I didn't use it.

Unfortunately, I wasn't able to come up with better solution, so I am open to ideas how to solve it. Because I know that this is not ideal.

Reply:
I agree with @clebergnu here, we should try to avoid adding this into the state machine and also that  tasks should create another tasks.

By looking at first glance, if the results are instances of `PostRuntimeTask`, this could be delegated to this class.

For sure you have thought about this more than us, and I would need to dig into the code with more time to understand better the limitations.